## 1. Load the data set

In [ ]:
#import necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns


%matplotlib inline
import matplotlib.pyplot as plt

from ydata_profiling import ProfileReport

In [4]:
#import the dataset
path = "dataset\credit_card.csv"
df = pd.read_csv(path)

<>:2: SyntaxWarning: invalid escape sequence '\c'
<>:2: SyntaxWarning: invalid escape sequence '\c'
C:\Users\kimwo\AppData\Local\Temp\ipykernel_24148\1620473683.py:2: SyntaxWarning: invalid escape sequence '\c'
  path = "dataset\credit_card.csv"


In [5]:
#check if the dataset is loaded 
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5     2 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6     4  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7     7 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8     7 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9     9 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

## 2. Pre-process and clean the dataset

In [ ]:
#converr the columns name to lower case
print("Columns before lower case conversion:",df.columns.tolist())
print("*"*50)
df.columns = df.columns.str.lower()
print("Columns after lower case conversion:", df.columns.tolist())

Columns before lower case conversion: ['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class']
**************************************************
Columns after lower case conversion: ['time', 'v1', 'v2', 'v3', 'v4', 'v5', 'v6', 'v7', 'v8', 'v9', 'v10', 'v11', 'v12', 'v13', 'v14', 'v15', 'v16', 'v17', 'v18', 'v19', 'v20', 'v21', 'v22', 'v23', 'v24', 'v25', 'v26', 'v27', 'v28', 'amount', 'class']


In [ ]:
#check the info of the dataset 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   time    999 non-null    int64  
 1   v1      999 non-null    float64
 2   v2      999 non-null    float64
 3   v3      999 non-null    float64
 4   v4      999 non-null    float64
 5   v5      999 non-null    float64
 6   v6      999 non-null    float64
 7   v7      999 non-null    float64
 8   v8      999 non-null    float64
 9   v9      999 non-null    float64
 10  v10     999 non-null    float64
 11  v11     999 non-null    float64
 12  v12     999 non-null    float64
 13  v13     999 non-null    float64
 14  v14     999 non-null    float64
 15  v15     999 non-null    float64
 16  v16     999 non-null    float64
 17  v17     999 non-null    float64
 18  v18     999 non-null    float64
 19  v19     999 non-null    float64
 20  v20     999 non-null    float64
 21  v21     999 non-null    float64
 22  v2

For security reasons, the feature names have been anonymized as V1, V2, and so on. Since the "time" column is not relevant in this case, we will drop it in the section below.

In [8]:
#drop the time column
df.drop(columns=['time'])

v1        v2        v3        v4        v5        v6        v7  \
0   -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1    1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
..        ...       ...       ...       ...       ...       ...       ...   
994  1.284612  0.303073 -0.067909  0.930684  0.117387 -0.487005  0.204739   
995 -0.654892  0.608319  1.585102 -3.009429  0.037593 -1.954023  1.335977   
996 -2.101171 -0.227365  1.624668 -0.291123  1.902446 -1.483921 -0.275117   
997 -1.248163  0.315246  3.708935  0.687280 -0.873071  1.091287  0.297707   
998  1.374134 -1.767210 -0.433352 -2.229552  0.331135  3.924775 -2.049947   

           v8        v9       v10  ...       v21       v22       v23  \
0    0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474   
1    0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288   
2    0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412   
3    0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321   
4   -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458   
..        ...       ...       ...  ...       ...       ...       ...   
994 -0.174873  0.058270  0.015465  ... -0.023272 -0.043358 -0.209701   
995 -0.612858  0.690254 -1.681537  ... -0.078527 -0.064194 -0.107350   
996  0.085964 -0.563098 -0.376024  ... -0.313782 -0.804784 -0.474101   
997 -0.633135  1.102317  2.536557  ... -0.824013  0.057907 -0.282351   
998  1.001403 -1.183310  1.278360  ... -0.252942 -0.461534 -0.030890   

          v24       v25       v26       v27       v28  amount  class  
0    0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1   -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2   -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3   -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4    0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
..        ...       ...       ...       ...       ...     ...    ...  
994 -0.446091  0.791745 -0.276971  0.009979  0.013268   14.05      0  
995  0.961776 -0.067760 -0.549465 -0.232588 -0.108261    3.90      0  
996  0.008102  0.259725  0.032376  0.323580  0.034622    1.78      0  
997  0.630774  0.283506 -0.204264  0.097555 -0.670480   30.00      0  
998  0.997119  0.384299 -0.187538  0.068817  0.038009   82.37      0  

[999 rows x 30 columns]

### Handle Missing Values

In [9]:
#check if any null or na values 
print("Check any null or na values in the dataset")
print("*"*50)
print("Check if column has any null\n", df.isnull().sum())

Check any null or na values in the dataset
**************************************************
Check if column has any null
 time      0
v1        0
v2        0
v3        0
v4        0
v5        0
v6        0
v7        0
v8        0
v9        0
v10       0
v11       0
v12       0
v13       0
v14       0
v15       0
v16       0
v17       0
v18       0
v19       0
v20       0
v21       0
v22       0
v23       0
v24       0
v25       0
v26       0
v27       0
v28       0
amount    0
class     0
dtype: int64


### Handle Duiplicate Values

In [10]:
#check duiplicate values in each column and print the count of the duplicates
for col in df.columns:
    duiplicate_count = df[col].duplicated().sum()
    print(f"Column:{col}")
    print(f"Duiplicate count:{duiplicate_count}")
    print("*"*50)

Column:time
Duiplicate count:459
**************************************************
Column:v1
Duiplicate count:8
**************************************************
Column:v2
Duiplicate count:8
**************************************************
Column:v3
Duiplicate count:8
**************************************************
Column:v4
Duiplicate count:8
**************************************************
Column:v5
Duiplicate count:8
**************************************************
Column:v6
Duiplicate count:8
**************************************************
Column:v7
Duiplicate count:8
**************************************************
Column:v8
Duiplicate count:8
**************************************************
Column:v9
Duiplicate count:8
**************************************************
Column:v10
Duiplicate count:8
**************************************************
Column:v11
Duiplicate count:8
**************************************************
Column:v12
Duiplicate count:8
****

In [11]:
#Displat the dataframe having missing values
df_missing_values = df[df.isnull().any(axis=1)]
df_missing_values.shape

(0, 31)

The above result shows that no missing values were found in any of the 31 columns, and the section below further confirms this.

In [12]:
df_missing_values.tail()

Empty DataFrame
Columns: [time, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11, v12, v13, v14, v15, v16, v17, v18, v19, v20, v21, v22, v23, v24, v25, v26, v27, v28, amount, class]
Index: []

[0 rows x 31 columns]

The result above means no missing values are found in the dataset 

### Handle outliners

In [ ]:
#define a function to find outliers using IQR method 
def find_outliers_IQR(input_df, feature):
    IQR = input_df[feature].quantile(0.75) - input_df[feature].quantile(0.25)
    lower_limit = input_df[feature].quantile(0.25) - (IQR * 1.5)
    upper_limit = input_df[feature].quantile(0.75) + (IQR * 1.5)
    return lower_limit, upper_limit

In [14]:
#find lower and upper for amount
lower_limit, upper_limit = find_outliers_IQR(df, 'amount')
print(f"Lower limit for amount: {lower_limit}")
print(f"Upper limit for amount: {upper_limit}")

Lower limit for amount: -69.25
Upper limit for amount: 129.91


In [15]:
#remove outliers from the dataset
df_cleaned = df[(df['amount']>lower_limit) & (df['amount']<upper_limit)]

print(f"Cleaned dataset shape:{df_cleaned.shape}")
print(f"Outliers count: {len(df)- len(df_cleaned)}")

Cleaned dataset shape:(886, 31)
Outliers count: 113


By taking this approach, we identified 113 outliers in the "amount" column. Before starting feature engineering, we assume that "amount" may be a critical feature in the model.

## 3. Exploratory Data Analysis (EDA)